paimon-core/src/main/java/org/apache/paimon/sort/zorder/ZIndexer.java

```diff
@@ -0,0 +1,373 @@
+import org.apache.paimon.types.LocalZonedTimestampType;
+import org.apache.paimon.types.MapType;
+import org.apache.paimon.types.MultisetType;
+import org.apache.paimon.types.RowType;
+import org.apache.paimon.types.SmallIntType;
+import org.apache.paimon.types.TimeType;
+import org.apache.paimon.types.TimestampType;
+import org.apache.paimon.types.TinyIntType;
+import org.apache.paimon.types.VarBinaryType;
+import org.apache.paimon.types.VarCharType;
+import org.apache.paimon.utils.ByteBuffers;
+import org.apache.paimon.utils.ZOrderByteUtils;
+
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+
+import static org.apache.paimon.utils.ZOrderByteUtils.PRIMITIVE_BUFFER_SIZE;
+
+/** Z-indexer for responsibility to generate z-index. */
+public class ZIndexer implements Serializable {
+
+    private final Set<ZorderBaseFunction> functionSet;
+    private final int[] fieldsIndex;
+    private final int totalBytes;
+    private transient ByteBuffer reuse;
+
+    public ZIndexer(RowType rowType, List<String> orderColumns) {
+        List<String> fields = rowType.getFieldNames();
+        fieldsIndex = new int[orderColumns.size()];
+        for (int i = 0; i < fieldsIndex.length; i++) {
+            int index = fields.indexOf(orderColumns.get(i));
+            if (index == -1) {
+                throw new IllegalArgumentException(
+                        "Can't find column: "
+                                + orderColumns.get(i)
+                                + " in row type fields: "
+                                + fields);
+            }
+            fieldsIndex[i] = index;
+        }
+        this.functionSet = constructFunctionMap(rowType.getFields());
+        this.totalBytes = PRIMITIVE_BUFFER_SIZE * this.fieldsIndex.length;
+    }
+
+    public void open() {
+        this.reuse = ByteBuffer.allocate(totalBytes);
+        functionSet.forEach(ZorderBaseFunction::open);
+    }
+
+    public int size() {
+        return totalBytes;
+    }
+
+    public byte[] index(InternalRow row) {
+        byte[][] columnBytes = new byte[fieldsIndex.length][];
+
+        int index = 0;
+        for (ZorderBaseFunction f : functionSet) {
+            columnBytes[index++] = f.apply(row);
+        }
+
+        return ZOrderByteUtils.interleaveBits(columnBytes, totalBytes, reuse);
+    }
+
+    public Set<ZorderBaseFunction> constructFunctionMap(List<DataField> fields) {
+        Set<ZorderBaseFunction> zorderFunctionSet = new LinkedHashSet<>();
+        // Construct zorderFunctionSet and fill dataTypes, rowFields
+        for (int fieldIndex = 0; fieldIndex < fieldsIndex.length; fieldIndex++) {
+            int index = fieldsIndex[fieldIndex];
+            DataField field = fields.get(index);
+            zorderFunctionSet.add(zmapColumnToCalculator(field).setPosition(index));
```

Comment:
remove `setPosition`.
You should pass index from constructor.

Reply:
done